### PR TITLE
Fix timeout and singleshot

### PR DESCRIPTION
**Pull request by OCopping**

### CODE REVIEW

**Comment by GDYendell** [comment]
Could you split this into two commits? The changes seem unrelated.

**Comment by OCopping** [comment]
I think this appears to work fine, can I merge it in?

File: src/pmacFilterControl.cpp [hunk 1]
```diff
@@ -491,9 +491,10 @@ void PMACFilterController::_process_state_changes() {
             this->_transition_state(ControlState::WAITING);
         }
     } else if (this->mode_ == ControlMode::SINGLESHOT) {
-        if (this->state_ == ControlState::IDLE || this->state_ == ControlState::WAITING) {
-            this->_transition_state(ControlState::SINGLESHOT_WAITING);
-        }
+        if ((this->state_ == ControlState::IDLE || this->state_ == ControlState::WAITING) && this->singleshot_start_ == false)
+            {
+                this->_transition_state(ControlState::SINGLESHOT_WAITING);
+            }
```
Comment (GDYendell) [inline src/pmacFilterControl.cpp:497]: The formatting is wrong here.

File: src/pmacFilterControl.cpp [hunk 2]
```diff
@@ -491,9 +491,10 @@ void PMACFilterController::_process_state_changes() {
             this->_transition_state(ControlState::WAITING);
         }
     } else if (this->mode_ == ControlMode::SINGLESHOT) {
-        if (this->state_ == ControlState::IDLE || this->state_ == ControlState::WAITING) {
-            this->_transition_state(ControlState::SINGLESHOT_WAITING);
-        }
+        if ((this->state_ == ControlState::IDLE || this->state_ == ControlState::WAITING) && this->singleshot_start_ == false)
```
Comment (GDYendell) [inline src/pmacFilterControl.cpp:494]:
```suggestion
        if ((this->state_ == ControlState::IDLE || this->state_ == ControlState::WAITING) && !this->singleshot_start_)
```

File: src/pmacFilterControl.cpp [hunk 3]
```diff
@@ -556,6 +556,9 @@ void PMACFilterController::_transition_state(ControlState state) {
                 _get_time(&this->last_process_ts_);
         }
         bool waiting = state == ControlState::WAITING || state == ControlState::SINGLESHOT_WAITING;
+        if ((this->singleshot_start_ == true && this->state_ == ControlState::WAITING) || state == ControlState::SINGLESHOT_COMPLETE) {
```
Comment (GDYendell) [inline src/pmacFilterControl.cpp:559]:
```suggestion
        if ((this->singleshot_start_ && this->state_ == ControlState::WAITING) || state == ControlState::SINGLESHOT_COMPLETE) {
```

**Review by GDYendell — COMMENTED** [review]
Could you add some inline comments to explain the logic? e.g. when transitioning from X to Y set / delay setting Z so that bad thing doesn't happen

File: src/pmacFilterControl.cpp [hunk 4]
```diff
@@ -556,7 +558,14 @@ void PMACFilterController::_transition_state(ControlState state) {
...
         bool waiting = state == ControlState::WAITING || state == ControlState::SINGLESHOT_WAITING;
-        if (state < 1 || (waiting && this->state_ >= 0)) {
+        // If currently in a scan and current state is waiting, or the scan is now complete, reset singleshot start flag
+        if (
+            (this->singleshot_start_ && this->state_ == ControlState::WAITING) || state == ControlState::SINGLESHOT_COMPLETE
+        ) {
+            this->singleshot_start_ = false;
+        }
+        if (state < 1 || (waiting && this->state_ >= 0)
+        ) {
```
Comment (GDYendell) [inline src/pmacFilterControl.cpp:568]:
```suggestion
        if (state < 1 || (waiting && this->state_ >= 0)) {
```

**Review by GDYendell — APPROVED** [review]

**Pull request merged**